docs/rearrange.md

```diff
@@ -0,0 +1,180 @@
+
+### Examples
+
+Examples are probably the best way to get a feel for the syntax:
+
+```python
+import haliax as hax
+import jax.random as jrandom
+
+N = hax.Axis("N", 32)
+C = hax.Axis("C", 3)
+H = hax.Axis("H", 64)
+W = hax.Axis("W", 64)
+
+x = hax.random.normal(jrandom.PRNGKey(0), (N, C, H, W))
+
+# transpose/permute axes
+y = hax.rearrange(x, "N C H W -> N H W C")
+# names don't have to match with positional syntax
+z = hax.rearrange(x, "num ch h w -> num h w ch")
+# ellipsis can be used to specify the rest of the dimensions
+z = hax.rearrange(x, "N C ... -> N ... C")
+
+# unordered syntax allows for select dimensions by name
+# transpose last two dimensions
+y = hax.rearrange(x, "{H W} -> ... W H")
+
+# don't know the order? use the unordered syntax
+y = hax.rearrange(x, "{W C H N} -> N H W C")
+
+# split dims as in einops
+y = hax.rearrange(x, "(step microbatch) ... -> step microbatch ...", step=4)
+# splitting dims can be done w/ unordered syntax
+y = hax.rearrange(x, "{(N: step microbatch) ...} -> step microbatch ...", step=4)
+
+# merging dims requires a name
+x = hax.rearrange(y, "step microbatch ... -> (N: step microbatch) ...")
+
+# some fancier examples
+
+# split into patches
+y = hax.rearrange(x, "N C (patch_h H) (patch_w W) -> N (P: patch_h patch_w) C H W", H=4, W=4)
+# unordered version
+y = hax.rearrange(x, "{(H: patch_h H) (W: patch_w W) C } -> ... (P: patch_h patch_w) C H W", H=4, W=4)
+
+# split into patches, then merge patches and channels
+y = hax.rearrange(x, "N C (patch_h H) (patch_w W) -> N (P: patch_h patch_w) (C: C H W)", H=4, W=4)
+# unordered version
+y = hax.rearrange(x, "{(H: patch_h H) (W: patch_w W) C } -> ... (P: patch_h patch_w) (C: C H W)", H=4, W=4)
+```
+
+### Bindings: Aliasing and Sizing
+
+In the above examples we used keyword arguments to give sizes to split axes, which is the same
+as in einops. However, we can also use bindings to alias axes. For example:
+
+```python
+# this produces the same result as the previous example
+y2 = hax.rearrange(x, "N C (patch_h foo) (patch_w bar) -> N (P: patch_h patch_w) (C: C foo bar)", foo=hax.Axis("H", 4), bar=hax.Axis("W", 4))
+assert y.axes == y2.axes
+```
+
+This example is a bit contrived, but it lets us use shorter or different names in the string which is occasionally useful.
+You can actually just use strings to specify aliases, which is sometimes useful as well.
+For instance, if we wanted "P" to actually be called "patch", but wanted to keep the short syntax, we could do:
+
+```python
+y3 = hax.rearrange(x, "N C (nh ph) (nw pw) -> N (P: nh nw) (C: C ph pw)", P="patch", pw=4, ph=4)
+```
+
+
+
+### Differences from einops
+
+As you may have noticed, there are some differences from einops:
+
+* Merged axes must have a name: `(C: C H W)` instead of `(C H W)`
+* The unordered syntax with `{ ... }` is new but hopefully not too confusing: select dimensions by name, and optionally split them using, e.g., `(N: step microbatch)`.
```

Comment:
```suggestion
* The unordered syntax with `{ ... }` is new but hopefully not too confusing
* Select dimensions by name, and optionally split them using `:` , e.g., `(N: step microbatch)`.
```

Let me know if I understand it incorrectly. This seems to be two different points.

Reply:
The use of ellipsis doesn't seem confusing to me at least (maybe the confusion happens the code implementation rather than reading it). It seems to be a useful thing to have.

Reply:
i guess i meant that it's either `{ N }` or `{ (N: x y) }` (with the latter being splitting). I'll rework